### PR TITLE
WIP: add talk functions

### PR DESCRIPTION
**Pull request by maxbachmann**
**This is still work in progress and untested**

the idea behind this is that normal talk sentences as used until now are not enough for some modules.
This enhancement allows the use of functions aswell.
e.g.
en.json
```
{
	"example1": ['string1'],
	"example2": ['string2']
}
```
de.json
```
{
	"example1": ['string1']
}
```
de.py
```
def example2(replace: list, shortReplyMode: bool, **kwargs) -> str:
    return 'fancy german grammar foo'
```

This way the concept gives an easy way to define functions to handle more complex answer handling (through kwargs it is possible to pass any additional parameters that might need to be passed). As shown above it is possible to use normal string mappings in one language and the function mapping in another language (because german grammar sucks a lot more often than english grammer xD)

### CODE REVIEW

File: core/voice/TalkManager.py [hunk 1]
```diff
@@ -44,17 +54,31 @@ def loadTalks(self, moduleToLoad: str = None):
...
 				continue
 
+			self._langData[moduleName] = dict()
 			for langTalkFile in langTalksMountpoint.iterdir():
 				lang = langTalkFile.stem
 				try:
-					self._langData.setdefault(moduleName, dict())[lang] = json.loads(langTalkFile.read_text())
+					if langTalkFile.suffix == '.py':
+						talkImport = importlib.import_module(f'modules.{moduleName}.talks.{lang}')
+						mappings = dict(getmembers(talkImport, isfunction))
```
Comment (Psychokiller1888) [inline core/voice/TalkManager.py]: Codacy Issue found: [Unused variable 'mappings'](https://app.codacy.com/app/ProjectAlice/ProjectAlice/pullRequest?prid=4626272)

**Comment by Psychokiller1888** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Complexity increasing per file
==============================
- core/voice/TalkManager.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/ProjectAlice/ProjectAlice/pullRequest?prid=4626272&bid=15509096)

File: core/voice/TalkManager.py [hunk 2]
```diff
@@ -65,21 +89,28 @@ def getTexts(self, module, talk, strType='default') -> list:
 		return arr
 
 
-	def chooseTalk(self, talk: str, module: str, activeLanguage: str, defaultLanguage: str, shortReplyMode: bool) -> str:
-		try:
-			talkData = self._langData[module][activeLanguage][talk]
+	def _selectSentence(self, talkData: dict, shortReplyMode: bool) -> Union[str, Callable]:
```
Comment (Psychokiller1888) [inline core/voice/TalkManager.py]: Codacy Issue found: [Method could be a function](https://app.codacy.com/app/ProjectAlice/ProjectAlice/pullRequest?prid=4626272)

File: core/base/model/Module.py [hunk 3]
```diff
@@ -413,13 +413,16 @@ def databaseInsert(self, tableName: str, query: str = None, values: dict = None)
...
-		if replace:
-			string = string.format(*replace)
+		if Callable(talk):
+			talk(replace=replace, **kwargs)
 
-		return string
+		if replace:
+			talk = talk.format(*replace)
+		return talk
+
```
Comment (Psychokiller1888) [inline core/base/model/Module.py]: Codacy Issue found: [Trailing whitespace](https://app.codacy.com/app/ProjectAlice/ProjectAlice/pullRequest?prid=4626272)

**Comment by Psychokiller1888** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- core/voice/TalkManager.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/ProjectAlice/ProjectAlice/pullRequest?prid=4626272&bid=15509096)

File: core/base/model/Module.py (hunk 3 above)
Comment (Psychokiller1888) [inline core/base/model/Module.py]: Codacy Issue found: [Trailing whitespace](https://app.codacy.com/app/ProjectAlice/ProjectAlice/pullRequest?prid=4626272)

File: core/voice/TalkManager.py (hunk 2 above)
Comment (Psychokiller1888) [inline core/voice/TalkManager.py]: Codacy Issue found: [Method could be a function](https://app.codacy.com/app/ProjectAlice/ProjectAlice/pullRequest?prid=4626272)

**Comment by Psychokiller1888** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- core/voice/TalkManager.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/ProjectAlice/ProjectAlice/pullRequest?prid=4626272&bid=15509096)

**Comment by Psychokiller1888** [comment]
Yeah, I like it, this would allow fixed terms to be translated also

**Comment by maxbachmann** [comment]
should the structure be:
```
talks
├─ en.json
├─ de.json
└─ de.py
```
or
```
talks
├─ en
 |   └─ en.json
└─ de
     ├─ de.json
     └─ de.py
```

**Comment by Psychokiller1888** [comment]
I'm sorry, did not see your reply. I guess the second one could be better if we drastically increase the amount of supported languages as we unlink snips

**Comment by maxbachmann** [comment]
Yeah I did see that Mycroft uses this structure, since they do not just allow to add the ISO639-1 codes like `en`, but they allow to specify the country aswell e.g. `en-US`. While we do not support something like this yet, we might in the future when there are more people that are adding translations.

**Comment by maxbachmann** [comment]
This will wait be something for alpha5

**Pull request closed without merging**